packages/eslint-plugin/src/rules/no-use-before-define.ts

```diff
@@ -92,6 +92,16 @@ function isOuterVariable(
   );
 }
 
+/**
+ * Checks whether or not a given reference is a type reference.
+ */
+function isTypeReference(reference: TSESLint.Scope.Reference): boolean {
+  return (
+    reference.isTypeReference ||
+    reference.identifier.parent?.parent?.type === AST_NODE_TYPES.TSTypeQuery
```

Comment:
almost! but this will only match one case.

We need to match all of these cases, and more:

- `typeof a`
- `typeof a.b`
- `typeof a.b.c`

Instead an easier way to do this would be with a loop.
Based on the defined types for `TSTypeQuery`, we know that the `exprName` will either be an `Identifier` or a `TSQualifiedName`.
https://github.com/typescript-eslint/typescript-eslint/blob/d7dc108580cdcb9890ac0539e7223aedbff4a0ed/packages/types/src/ts-estree.ts#L1669-L1672

Based on the defined types for `TSQualifiedName`, we know the left will always be an `Identifier` or a `TSQualifiedName`.
https://github.com/typescript-eslint/typescript-eslint/blob/d7dc108580cdcb9890ac0539e7223aedbff4a0ed/packages/types/src/ts-estree.ts#L1572-L1576

This means we can simply loop and look at the parents
The reason I gave you the above is it'll help you constrain the loop so it won't iterate forever

Reply:
Fixed!